### ISSUE TITLE
The whole database schema is visible in client-side when using NextJS

### ISSUE THREAD

**Issue by juhawilppu** [prisma/prisma#22045]
## Problem

When using NextJS, all the autogenerated prisma model TypeScript types are visible to the client-side. Effectively this exposes the whole database schema to client-side. There are many tables and fields that should not visible to the clients, like log tables. In general, the backend implementation and how the data is stored, should not be visible to clients.

No data gets leaked, only the schema.

This relates to fields such as
- encryptedSocialSecurityNumber
- encryptedPhoneNumber
- encryptedEmail

## Suggested solution

Add a `@@visibility` setting to models and fields. Don't generate client-side types when visibility is set to `backend`.

This solution is similar to https://github.com/prisma/prisma/issues/5042 except it applies to the type definitions.

```
model PatientInformation {
  id                    Int         @id @default(autoincrement())
  encryptedSocialSecurityNumber String
  encryptedPhoneNumber  String
  encryptedEmail String
  @@map("patient_information")
  @@visibility("backend")
}
```

## Alternatives

Workaround: Somewhat possible to manually remove the type definitions from the minified JS chunks before installing to production.

## Additional context

Screenshot of the problem:
<img width="1605" alt="Screenshot 2023-11-21 at 9 17 11" src="https://github.com/prisma/prisma/assets/1705197/1988e81a-8b21-4558-acde-225080a8526d">

**Comment by SevInf** [comment]
Hi @juhawilppu.
Typescript types should be stripped before anything gets to the browser - not just for Prisma + Next.js but pretty much for any TS project. Could you provide a small reproduction where we could see it happening?

**Comment by ferka123** [comment]
you should probably use import type instead of import

**Comment by Yohannfra** [comment]
It might also be a good idea to enable this eslint rule

https://typescript-eslint.io/rules/consistent-type-imports/

**Comment by dvins** [comment]
Just curious, your but the backend for fronted is directly manipulating the data store via Prisma? You don't have an intermediary API be REST/GraphQL, etc. that abstracts (and protects) these operations?

**Comment by maiconcarraro** [comment]
> and then open browser to `http://localhost:3000/dynamic`.

@juhawilppu there is no dynamic route in the repo above, I tried to reproduce, but no success, did you face the same issue for production? you can try running start & build locally to emulate, curious about this

**Comment by juan-carlos-correa** [comment]
> ## Problem
> When using NextJS, all the autogenerated prisma model TypeScript types are visible to the client-side. Effectively this exposes the whole database schema to client-side. There are many tables and fields that should not visible to the clients, like log tables. In general, the backend implementation and how the data is stored, should not be visible to clients.
> 
> No data gets leaked, only the schema.
> 
> This relates to fields such as
> 
> * encryptedSocialSecurityNumber
> * encryptedPhoneNumber
> * encryptedEmail
> 
> ## Suggested solution
> Add a `@@visibility` setting to models and fields. Don't generate client-side types when visibility is set to `backend`.
> 
> This solution is similar to #5042 except it applies to the type definitions.
> 
> ```
> model PatientInformation {
>   id                    Int         @id @default(autoincrement())
>   encryptedSocialSecurityNumber String
>   encryptedPhoneNumber  String
>   encryptedEmail String
>   @@map("patient_information")
>   @@visibility("backend")
> }
> ```
> 
> ## Additional context
> Screenshot of the problem: <img alt="Screenshot 2023-11-21 at 9 17 11" width="1605" src="https://private-user-images.githubusercontent.com/1705197/284494558-1988e81a-8b21-4558-acde-225080a8526d.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjIxMzcwMDQsIm5iZiI6MTcyMjEzNjcwNCwicGF0aCI6Ii8xNzA1MTk3LzI4NDQ5NDU1OC0xOTg4ZTgxYS04YjIxLTQ1NTgtYWNkZS0yMjUwODBhODUyNmQucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDcyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA3MjhUMDMxODI0WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NzM1YWJjM2FlYmYxNDUwYTE5OGFjZTAzNzk4ODczODMzNzQ0ZjNkZGM2NWVmY2IwNmZiMmMyODNmMGEwNjYxMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.6VTmcn4SrYNmUSeunSNHBq8rpyQTw2yjpX1XxUbPlOc">

I can confirm that this is happening to me too.

My current WIP approach is to remove all the Prisma client imports from client components, and validate that the Prisma schemas are not exposed as part of the client bundled code.

I'm using Prisma in server actions, on my mind, that code should not be part of the client/frontend bundled.

**Comment by yunsii** [comment]
> you should probably use import type instead of import

How about when we need to import Enum?

**Comment by ferka123** [comment]
> > you should probably use import type instead of import
> 
> How about when we need to import Enum?

you could import a union instead 
import type {$Enums} from @prisma/client
$Enums.YourEnum as string | string | string

**Comment by juan-carlos-correa** [comment]
> > you should probably use import type instead of import
> 
> How about when we need to import Enum?

I ended up creating a new Enum 😔 

You could try the ferka123's approach, I think if you do an import of prisma client (even for making an union), the schema is going to be exposed anyway 🤔

**Comment by piyanggoon** [comment]
Same issue use next.js server action + prisma schema got leak on bundle.

**Comment by juan-carlos-correa** [comment]
> Same issue use next.js server action, prisma schema got leak on bundle.

Using a server action? do you have a code example? 

I'm using server actions too and don't have the prisma schema leaked on the bundle.

In theory, a server action is like a lambda or a serverless function that the frontend communicates via HTTP requests, so it should not be part of the bundled code in the browser 🤔

**Comment by piyanggoon** [comment]
> > Same issue use next.js server action, prisma schema got leak on bundle.
> 
> Using a server action? do you have a code example?
> 
> I'm using server actions too and don't have the prisma schema leaked on the bundle.
> 
> In theory, a server action is like a lambda or a serverless function that the frontend communicates via HTTP requests, so it should not be part of the bundled code in the browser 🤔

use on lib next-safe-action (https://next-safe-action.dev/).

**Comment by yunsii** [comment]
> > > you should probably use import type instead of import
> > 
> > 
> > How about when we need to import Enum?
> 
> you could import a union instead import type {$Enums} from @prisma/client $Enums.YourEnum as string | string | string

Yes, I got it, we can use $Enums directly in server side. But when used in client side, import type of $Enums instead. It seems we can import type always without caring about the import difference.
